x/wasm/IBC.md

```diff
@@ -0,0 +1,128 @@
+  an IBCEnabled contract, this will be routed to the proper contract
+  and call an exposed function for this purpose.
+* Handle `IBCAckPacket` - The original sender of `IBCSendMsg` will
+  get this callback eventually if the message was successfully
+  processed on the other chain
+* Handle `IBCErrorPacket` - The original sender of `IBCSendMsg` will 
+  get this callback eventually if the message failed to be
+  processed on the other chain (for any reason)
+
+For mocks, all the `IBCxxxPacket` commands are routed to some
+Golang stub handler, but containing the contract address, so we
+can perform contract-specific actions for each packet.
+
+There are open questions on how exactly the various `IBCxxPacket`
+messages are routed to a contract, as well as related to the
+channel lifecycle. (Can a contract open a channel? How do we
+allow contracts to participate in protocol version negotiation
+when establishing a channel?)
+
+## Ports and Channels
+
+There are two main proposals on how to map these concepts to
+contracts. We will describe both in detail, with arguments
+pro and contra, to select one approach.
+
+### One Port per Contract
+
+This may be the most straight-forward mapping, treating each contract 
+like a module. It does lead to very long portIDs however. Make special
+attention to both the Channel establishment (which should be compatible
+with standard ICS20 modules without changes on their part), as well
+as how contracts can properly identify their counterparty.
+
+* Upon `Instantiate`, if a contract is *IBC Enabled*, we dynamically 
+  bind a port for this contract. The port name is `wasm-<contract address>`,
+  eg. `wasm-cosmos1hmdudppzceg27qsuq707tjg8rkgj7g5hnvnw29`
+* If a *Channel* is being established with a registered `wasm-xyz` port,
+  the `x/wasm.Keeper` will handle this and call into the appropriate
+  contract to determine supported protocol versions during the
+  [`ChanOpenTry` and `ChanOpenAck` phases](https://docs.cosmos.network/master/ibc/overview.html#channels).
+  (See [Channel Handshake Version Negotiation](https://docs.cosmos.network/master/ibc/custom.html#channel-handshake-version-negotiation))
+* Both the *Port* and the *Channel* are fully owned by one contract.
+* `x/wasm` only accepts *ORDERED Channels* for simplicity of contract
+  correctness.
+* When sending a packet, the CosmWasm contract must specify the *ChannelID*
+  (and remote *PortID*?), which is generally provided by the external client
+  which understands which channels it wishes to communicate over.
+* When sending a packet, the contract can set a custom identifier (of
+  max 64 characters) that it can use to look this up later.
+* When receiving a Packet (or Ack or Error), the contracts receives the
+  *ChannelID* as well as remote *PortID* (and *ClientID*???) that is
+  communicating.
+* When receiving an Ack or Error packet, the contract also receives the
+  same identifier that it set on Send (`x/wasm` handles mapping between this
```

Comment:
I would prefer to pass the origin message payload here. It is available in the IBC packets. This would leave it to the contract to store and pass any identifier.